File: frontend/src/api/portfolio.ts
import apiClient from './client';

export interface TransactionCreate {
  asset_type: 'stock' | 'mutual_fund' | 'emergency_fund' | 'fixed_deposit' | 'bond';
  symbol: string;
  name: string;
  transaction_type: 'BUY' | 'SELL' | 'SIP' | 'DIVIDEND' | 'BONUS' | 'SPLIT';
  transaction_date: string; // YYYY-MM-DD
  quantity: number;
  price: number;
  brokerage?: number;
  taxes?: number;
  notes?: string;
}

export interface TransactionResponse extends TransactionCreate {
  id: number;
  user_id: number;
  amount: number;
  created_at: string;
  updated_at: string;
}

export interface HoldingSummary {
  symbol: string;
  name: string;
  asset_type: 'stock' | 'mutual_fund' | 'emergency_fund' | 'fixed_deposit' | 'bond';
  quantity: number;
  avg_price: number;
  invested_amount: number;
  current_price: number;
  current_value: number;
  unrealized_pnl: number;
  unrealized_pnl_percent: number;
  day_change?: number;
  day_change_percent?: number;
}

export interface AssetAllocation {
  stocks_value: number;
  stocks_invested: number;
  stocks_count: number;
  mf_value: number;
  mf_invested: number;
  mf_count: number;
  emergency_fund_value: number;
  emergency_fund_invested: number;
  emergency_fund_count: number;
  fixed_deposit_value: number;
  fixed_deposit_invested: number;
  fixed_deposit_count: number;
  bond_value: number;
  bond_invested: number;
  bond_count: number;
  stocks_percentage: number;
  mf_percentage: number;
  emergency_fund_percentage: number;
  fixed_deposit_percentage: number;
  bond_percentage: number;
}

export interface PortfolioSummary {
  total_invested: number;
  current_value: number;
  total_pnl: number;
  total_pnl_percent: number;
  day_change: number;
  day_change_percent: number;
  xirr?: number;
  allocation: AssetAllocation;
  holdings: HoldingSummary[];
}

export async function addTransaction(payload: TransactionCreate): Promise<TransactionResponse> {
  const response = await apiClient.post<TransactionResponse>('/portfolio/transactions', payload);
  return response.data;
}

export async function getTransactions(
  assetType?: 'stock' | 'mutual_fund' | 'emergency_fund' | 'fixed_deposit' | 'bond',
  symbol?: string
): Promise<TransactionResponse[]> {
  const response = await apiClient.get<TransactionResponse[]>('/portfolio/transactions', {
    params: { asset_type: assetType, symbol },
  });
  return response.data;
}

export async function deleteTransaction(txnId: number): Promise<void> {
  await apiClient.delete(`/portfolio/transactions/${txnId}`);
}

export async function getPortfolioSummary(): Promise<PortfolioSummary> {
  const response = await apiClient.get<PortfolioSummary>('/portfolio/summary');
  return response.data;
}

File: backend/app/schemas/portfolio.py
"""
Portfolio Pydantic Schemas
"""

from datetime import date, datetime
from typing import Optional, List, Dict
from pydantic import BaseModel, Field, ConfigDict


class TransactionBase(BaseModel):
    asset_type: str = Field(..., description="'stock' or 'mutual_fund'")
    symbol: str = Field(..., description="Stock symbol (e.g. RELIANCE) or MF scheme code (e.g. 122639)")
    name: str = Field(..., description="Display name of the asset")
    transaction_type: str = Field(..., description="BUY | SELL | SIP | DIVIDEND | BONUS | SPLIT")
    transaction_date: date = Field(..., description="Date of transaction")
    quantity: float = Field(..., gt=0, description="Quantity or units")
    price: float = Field(..., ge=0, description="Price per share or NAV per unit")
    brokerage: float = Field(0.0, ge=0, description="Brokerage charges")
    taxes: float = Field(0.0, ge=0, description="STT / Taxes")
    notes: Optional[str] = Field(None, max_length=1000)


class TransactionCreate(TransactionBase):
    pass


class TransactionResponse(TransactionBase):
    id: int
    user_id: int
    amount: float
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class HoldingSummary(BaseModel):
    symbol: str
    name: str
    asset_type: str  # 'stock' | 'mutual_fund'
    quantity: float
    avg_price: float
    invested_amount: float
    current_price: float
    current_value: float
    unrealized_pnl: float
    unrealized_pnl_percent: float
    day_change: Optional[float] = 0.0
    day_change_percent: Optional[float] = 0.0


class AssetAllocation(BaseModel):
    stocks_value: float = 0.0
    stocks_invested: float = 0.0
    stocks_count: int = 0
    mf_value: float = 0.0
    mf_invested: float = 0.0
    mf_count: int = 0
    emergency_fund_value: float = 0.0
    emergency_fund_invested: float = 0.0
    emergency_fund_count: int = 0
    fixed_deposit_value: float = 0.0
    fixed_deposit_invested: float = 0.0
    fixed_deposit_count: int = 0
    bond_value: float = 0.0
    bond_invested: float = 0.0
    bond_count: int = 0
    stocks_percentage: float = 0.0
    mf_percentage: float = 0.0
    emergency_fund_percentage: float = 0.0
    fixed_deposit_percentage: float = 0.0
    bond_percentage: float = 0.0


class PortfolioSummary(BaseModel):
    total_invested: float = 0.0
    current_value: float = 0.0
    total_pnl: float = 0.0
    total_pnl_percent: float = 0.0
    day_change: float = 0.0
    day_change_percent: float = 0.0
    xirr: Optional[float] = 0.0
    allocation: AssetAllocation
    holdings: List[HoldingSummary] = []
